src/VisualStudio/Core/Def/Implementation/AnalyzerDependency/AnalyzerDependencyCheckingService.cs

```diff
@@ -19,79 +18,131 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation
             category: FeaturesResources.Roslyn_HostError,
             defaultSeverity: DiagnosticSeverity.Warning,
             isEnabledByDefault: true);
 
-        private readonly DiagnosticDescriptor _analyzerDependencyConflictRule = new DiagnosticDescriptor(
+        private static readonly DiagnosticDescriptor s_analyzerDependencyConflictRule = new DiagnosticDescriptor(
             id: IDEDiagnosticIds.AnalyzerDependencyConflictId,
             title: ServicesVSResources.AnalyzerDependencyConflict,
             messageFormat: ServicesVSResources.Analyzer_assemblies_0_and_1_both_have_identity_2_but_different_contents_Only_one_will_be_loaded_and_analyzers_using_these_assemblies_may_not_run_correctly,
             category: FeaturesResources.Roslyn_HostError,
             defaultSeverity: DiagnosticSeverity.Warning,
             isEnabledByDefault: true);
 
+
         [ImportingConstructor]
         public AnalyzerDependencyCheckingService(
-            VisualStudioWorkspaceImpl workspace,
-            HostDiagnosticUpdateSource updateSource)
+            VisualStudioWorkspace workspace,
+            HostDiagnosticUpdateSource hostDiagnosticUpdateSource)
         {
             _workspace = workspace;
-            _updateSource = updateSource;
-            _bindingRedirectionService = new BindingRedirectionService();
+            _hostDiagnosticUpdateSource = hostDiagnosticUpdateSource;
         }
 
-        public async void CheckForConflictsAsync()
+        public void ReanalyzeSolutionForConflicts()
         {
-            AnalyzerDependencyResults results = null;
-            try
-            {
-                results = await GetConflictsAsync().ConfigureAwait(continueOnCapturedContext: true);
-            }
-            catch
-            {
-                return;
-            }
+            var solution = _workspace.CurrentSolution;
+            var currentAnalyzerPaths = solution
+                .Projects
+                .SelectMany(p => p.AnalyzerReferences)
+                .OfType<AnalyzerFileReference>()
+                .Select(a => a.FullPath)
+                .ToImmutableHashSet(StringComparer.OrdinalIgnoreCase);
 
-            if (results == null)
+            lock (_gate)
             {
-                return;
+                // If we've already started analysis for this set of analyzers, no reason to start over
+                if (currentAnalyzerPaths.SetEquals(_previousAnalyzerPaths))
+                {
+                    return;
+                }
+
+                _cancellationTokenSource.Cancel();
+                _cancellationTokenSource = new CancellationTokenSource();
+                _previousAnalyzerPaths = currentAnalyzerPaths;
+
+                // We are explicitly relying on SafeContinueWith including LazyCancellation as a continuation option here
```

Comment:
When the lazy cancellation issue was originally discovered, ContinueWith was outright banned from the codebase and everybody *had* to use SafeContinueWith, which was written prior to the LazyCancellation flag existing in TPL. The enforcement has since died (it was written prior to analyzers being a thing, and worked via IL decompilation if I recall), and LazyCancellation is now in the TPL so there's no reason anybody who actually wanted that behavior could just specify it.

One day in my copious free time I'd love to delete SafeContinueWith and just let people who want LazyCancellation use it rather than picking a policy. In some cases, eager cancellation might actually be preferrable in certain perf cases.

Reply:
It doesn't help that SafeContinueWith then also got error handling behaviors, so the original intent now is kinda lost.